technical/binary_tree/c++/binary_tree.h

```diff
@@ -0,0 +1,272 @@
+#include <stack>
+#include <iostream>
+
+template <typename T>
+// Expects type T to have an operator overload for <. 
+class binary_tree {
+  friend class iterator;
+private:
+  struct node {
+    T data;
+    node* left;
+    node* right;
+    node* parent;
+    node(T val, node* parent): data(val), left(nullptr), right(nullptr), parent(parent) {}
+  };
+  node* root;
+  void delete_helper(node*);
+  void add_helper(T, node* &);
+  bool find_helper(T, node*);
+  void print_helper(node*);
+  node*& find_node(T, node*&);
+  node* in_order_successor(node*);
+  node* in_order_predecessor(node*);
+  void remove_helper(node*&);
+public:
+  // does an in-order traversal of the tree
+  // modifying the tree invalidates the iterator
+  class iterator {
+    friend class binary_tree;
+  private:
+    std::stack<node*> nodes;
+  public:
+    iterator();
+    iterator(const iterator&); //cpy-ctor
+    iterator operator=(const iterator&);
+    iterator operator++();
+    T& operator*(); // get data at this node
+    bool operator!=(const iterator&);
+    bool operator==(const iterator&);
+  };
+  binary_tree();
+  ~binary_tree();
+  void add(T);
+  bool find(T);
+  bool empty();
+  iterator first();
+  iterator end();
+  void print();
+  void remove(T);
+};
+
+template <typename T>
+binary_tree<T>::binary_tree(): root(nullptr) {}
+
+template <typename T>
+void binary_tree<T>::delete_helper(binary_tree<T>::node* n) {
+  if(!n) return;
```

Comment:
So is delete_helper supposed to delete the entire sub tree? More comments explaining purpose of functions would be nice.